src/hooks/useEventListener/useEventListener.ts

```diff
@@ -0,0 +1,113 @@
+import { RefObject, useEffect } from 'react';
+
+import { usePreservedCallback } from '../usePreservedCallback/index.ts';
+
+/**
+ * @description
+ * `useEventListener` is a React hook that allows you to easily add and clean up event listeners on various targets,
+ * such as `window`, `document`, HTML elements, or SVG elements.
+ * The listener is automatically updated with the latest handler on each render without reattaching,
+ * ensuring stable performance and correct behavior.
+ *
+ * @template KW - Event name type for `window` events, determining the corresponding event object type.
+ * @template KD - Event name type for `document` events, determining the corresponding event object type.
+ * @template KH - Event name type for HTML or SVG element events, determining the corresponding event object type.
+ * @template T - Type of the DOM element being referenced (default is `HTMLElement`, but can be an SVG element).
+ * @param {KW | KD | KH} eventName - The name of the event to listen for.
+ * @param {(event: WindowEventMap[KW] | DocumentEventMap[KD] | HTMLElementEventMap[KH] | SVGElementEventMap[KH]) => void} handler - The callback function that will be triggered when the event occurs.
+ * @param {RefObject<T | null> | Document} [element] - The target to attach the event listener to. Can be a React `ref` object or the `document`. If omitted or `undefined`, the listener is attached to the `window`.
+ * @param {boolean | AddEventListenerOptions} [options] - Optional parameters for the event listener such as `capture`, `once`, or `passive`.
+ *
+ * @example
+ * function WindowResize() {
+ *   useEventListener('resize', (event) => {
+ *     console.log('Window resized', event);
+ *   });
+ *
+ *   return <div>Resize the window and check the console.</div>;
+ * }
+ *
+ * @example
+ * function ClickButton() {
+ *   const buttonRef = useRef<HTMLButtonElement>(null);
+ *
+ *   useEventListener('click', (event) => {
+ *     console.log('Button clicked', event);
+ *   }, buttonRef);
+ *
+ *   return <button ref={buttonRef}>Click me</button>;
+ * }
+ *
+ * @example
+ * function ScrollTracker() {
+ *   const scrollRef = useRef<HTMLDivElement>(null);
+ *
+ *   useEventListener('scroll', () => {
+ *     console.log('Scroll event detected!');
+ *   }, scrollRef, { passive: true });
+ *
+ *   return (
+ *     <div ref={scrollRef} style={{ height: '100px', overflowY: 'scroll' }}>
+ *       <div style={{ height: '300px' }}>Scroll Me!</div>
+ *     </div>
+ *   );
+ * }
+ *
+ * @example
+ * function Document() {
+ *   useEventListener('click', (event) => {
+ *     console.log('Document clicked at coordinates', event.clientX, event.clientY);
+ *   }, document);
```

Comment:
This implementation does not appear to be safe in an SSR (Server-Side Rendering) environment. Directly referencing client-side global variables such as document seems risky. We should look for a way to ensure safe operation even in SSR environments.

For example, when you need to attach handlers to client-side global variables like `document` or `window`, instead of referencing them directly, you can pass them as strings like `'document'` or `'window'`. Then, inside a `useEffect`, you can safely access them conditionally. This approach can help ensure stability in SSR environments.

```suggestion
 *   useEventListener('click', (event) => {
 *     console.log('Document clicked at coordinates', event.clientX, event.clientY);
 *   }, 
 *   // ensure that `document` is only referenced on the client side.
 *   'document'
 * );
```